python/pyarrow/tests/test_dataset.py

```diff
@@ -5084,10 +5084,8 @@ def test_dataset_partition_with_slash(tmpdir):
     read_table = ds.dataset(
         source=path,
         format='ipc',
-        partitioning='hive',
-        schema=pa.schema([pa.field("exp_id", pa.int32()),
-                          pa.field("exp_meta", pa.utf8())])
-    ).to_table().combine_chunks()
+        schema=dt_table.schema,
```

Comment:
Does the schema even need to be specified here?

Reply:
Not at all, but there was a bug when the schema was specified.

Reply:
To evaluate that, I left it as it is but simplified it.

Reply:
I am a trying to gauge the diff here, but did something actually change? It seems that the schema that was specified manually before should be identical to `dt_table.schema`, and so this change is just a simplification? 
But what is then the "bug when the schema was specified" you are mentioning?

Reply:
In a previous PR I added this test case. It works without providing the schema (with the current change) and we don't need the change I did in the `_ensure_scheme`. But if a user provide a schema in this method, I get nulls for the `exp_meta` column after making this change. That was the bug encountered. I wasn't quite confident changing the test case to make things work. Do you agree with the C++ change to fix this bug? If so, for Python we could just leave the `schema`, but in case someone provides it there, we could run into trouble.

Reply:
> It works without providing the schema

But a schema_already_ was specified before this PR? I don't understand what this PR is changing about this test. 

It might be clearer if you add a new test that fails on current master, and is fixed by this PR.

Reply:
I have added a new test to evaluate the failing idea in the `test_dataset.py`, but an existing test case fails with the current modification.

Reply:
The issue is with the existing test case. When the schema is provided that case failed, but I realized the fix was to handle it where the `_ensure_scheme` is defined.

Reply:
I could be wrong here. Any thoughts?

Reply:
> The issue is with the existing test case. When the schema is provided that case failed

Can you show the error that you got for this test before you made the changes to `_ensure_partitioning`?

(and to be clear, the test also fails without the modifications to the test code? As those changes are only stylistic?)

Reply:
Ah it is simple. Basically the data in the `exp_meta` is all `null` values, it shouldn't be the case. 
What I observed was, when we create the scheme, we should pass the schema in that case. Maybe that is a wrong fix.

Reply:
I am also confused.  Here is my understanding of this test:

 * A user wants to partition on column exp_meta
 * The values in exp_meta contain slashes
 * We should be able to partition on that column and still write and then read the dataset

The files that will get created will be of the shape:

```
<tmpdir>/slash-writer-x/exp_meta=experiment%2FA%2Ff.csv/chunk-0.parquet
```

There should be no need to specify a schema manually to read this dataset.

It should also be harmless to specify a schema.  However, doing so makes the test less clear.

Reply:
I agree, we don't have to provide the schema. In the PR we added this, I have included the schema. I think this could be my misunderstanding. 

But my concern is, if one does specify that schema, with this new change, we are getting null values in the `exp_meta` column.

Reply:
In theory we don't need to pass the schema here, but if one does, we get invalid results. I am not quite sure which one to change :thinking:

Reply:
> In theory we don't need to pass the schema here, but if one does, we get invalid results

In the current code?  Or in this PR?  Or both?

Both passing and not passing the schema should work.  If passing the schema gives invalid results then I would ask we follow @jorisvandenbossche 's advice and create a test that isolates this behavior as much as possible.  This may require further debugging to figure out exactly what happened.

Reply:
Yes, I will further debug and see how to replicate this with an isolated test.

Reply:
I am going to remove my review request until we get this sorted out, just for my own book-keeping.  Feel free to re-request as soon as this is ready for another look.